src/tests/ftest/util/yaml_utils.py

```diff
@@ -0,0 +1,485 @@
+            devices = auto_detect_devices(host_list, "VMD", "4", None)
+        if devices:
+            device_types.append("VMD")
+
+    # Second check for any non-VMD NVMe disks, if requested
+    if nvme_args[0] in ["auto", "auto_nvme"]:
+        dev_list = auto_detect_devices(host_list, "NVMe", "4", dev_filter)
+        if dev_list:
+            devices.extend(dev_list)
+            device_types.append("NVMe")
+
+    # If no VMD or NVMe devices were found raise an exception
+    if not devices:
+        raise CommandFailure(
+            "Unable to auto-detect devices for the '--nvme {}' argument".format(args.nvme))
+
+    log.info(
+        "Auto-detected %s devices on %s: %s", " & ".join(device_types), args.test_servers, devices)
+    return ",".join(devices)
+
+
+def auto_detect_devices(host_list, device_type, length, device_filter=None):
+    """Get a list of NVMe/VMD devices found on each specified host.
+
+    Args:
+        host_list (NodeSet): hosts on which to find the NVMe/VMD devices
+        device_type (str): device type to find, e.g. 'NVMe' or 'VMD'
+        length (str): number of digits to match in the first PCI domain number
+        device_filter (str, optional): optional filter to apply to device searching. Defaults to
+            None.
+
+    Raises:
+        CommandFailure: if an invalid device_type is specified
+
+    Returns:
+        list: A list of detected devices - empty if none found
+
+    """
+    log = getLogger()
+    devices = []
+
+    # Find the devices on each host
+    if device_type == "VMD":
+        # Exclude the controller revision as this causes heterogeneous clush output
+        command_list = [
+            "/sbin/lspci -D",
+            "grep -E '^[0-9a-f]{{{0}}}:[0-9a-f]{{2}}:[0-9a-f]{{2}}.[0-9a-f] '".format(length),
+            "grep -E 'Volume Management Device NVMe RAID Controller'",
+            r"sed -E 's/\(rev\s+([a-f0-9])+\)//I'"]
+    elif device_type == "NVMe":
+        command_list = [
+            "/sbin/lspci -D",
+            "grep -E '^[0-9a-f]{{{0}}}:[0-9a-f]{{2}}:[0-9a-f]{{2}}.[0-9a-f] '".format(length),
+            "grep -E 'Non-Volatile memory controller:'"]
+        if device_filter and device_filter.startswith("-"):
+            command_list.append("grep -v '{}'".format(device_filter[1:]))
+        elif device_filter:
+            command_list.append("grep '{}'".format(device_filter))
+    else:
+        raise CommandFailure(
+            "Invalid 'device_type' for NVMe/VMD auto-detection: {}".format(device_type))
+
+    command = " | ".join(command_list) + " || :"
+    task = run_task(host_list, command, verbose=True)
+
+    # Verify the command was successful on each server host
+    if display_task(task):
+        # Verify each server host has the same VMD PCI addresses
+        output_data = list(task.iter_buffers())
+        if len(output_data) > 1:
+            log.error("ERROR: Non-homogeneous {} PCI addresses.".format(device_type))
```

Comment:
(pylint-logging-format-interpolation) Use lazy % formatting in logging functions